hcloud/core/domain.py

```diff
@@ -16,6 +18,15 @@ def __repr__(self) -> str:
         kwargs = [f"{key}={getattr(self, key)!r}" for key in self.__api_properties__]  # type: ignore[var-annotated]
         return f"{self.__class__.__qualname__}({', '.join(kwargs)})"
 
+    def __eq__(self, other: Any) -> bool:
+        """Compare a domain object with another of the same type."""
+        if type(self) != type(other):
+            return False
+        for key in self.__api_properties__:
+            if getattr(self, key) != getattr(other, key):
+                return False
+        return True
```

Comment:
The domain classes are then mixed into the `Bound<Resource>` classes, which might sometimes be incomplete.

When getting a value from an incomplete bound resource, the model is magically fetched from the API. I am wondering if this might cause a lot of magical API requests without the user knowing. Should the equality function on the bound model be reset to the base `object` equality function ?

Reply:
Hrm, in that case maybe the bound model should override this, and compare data_models?

I had also wondered if we should allow comparing subclasses rather than the strict type check. This bound resource issue is probably a good argument for keeping the check strict.

Reply:
I think its is best to override the bound model equality to check the data model indeed.

Comparing a bound model to a model should be supported too, or we might introduce confusing behaviors.

Reply:
> Comparing a bound model to a model should be supported too, or we might introduce confusing behaviours.

I can't imagine how to do that sanely, because we can't know if attributes on an un-bound model differ from the attributes known on the server side.

Do you have an algorithm in mind?

Reply:
I've ended up with something that does successfully compare bound and un-bound models. Things will not be considered equal if they have different amounts of data present.